README.md
<div id="top"></div>


<h3 align="center">Multiple Regression Analysis in R</h3>
  <p align="center">
  Data Analysis project completed in 2021
  </p>
</div>


<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#acknowledgments">Acknowledgments</a></li>
  </ol>
</details>


## About The Project

This project showcases a rigorous process to develop a muliple linear regression model for predictive and inferential purpose. 
It uses a dataset of 1000 properties in California, with 14 variables, including the identifier and response. The goal is to understand the factors that affect a property's
housing price, and to make good prediction on the housing price given some new data with other information. The analytical process focuses on data description, 
data visualization, variable selection, model transformation, model diagnostics, model validation, and model interpretation. The concepts used include ANOVA, power 
transformation, analysis of covariance, AIC(Akaike Information Criteria), ajusted AIC, BIC, QQ plots, stepwise variable selection, VIF(variance inflation factor). The two 
preconditions and four assumptions for a multiple linear regression model are checked and the model is handled appropriately to deal with any violation. In the analysis of 
covariance, models with different main effect and interaction terms are investigated and compared. The variables have been checked for multicollinearity, and highly correlated 
variables and high leverage points are removed after taking the context into consideration.

### Built With
R Libraries:
* car
* MASS
* leaps
* pnag
* xtable

### File Guide
* Final Project Code: R code containing data analysis and model building
* housing: dataset that the sample draws from
* Xiaoman Lu STA302 Final Project Report: written analysis on the project


<p align="right">(<a href="#top">back to top</a>)</p>

<!-- USAGE EXAMPLES -->
## Usage

The procedure used in this project can be referenced for applying multiple linear regression in other datasets. Please note that it cannot replace a through conceptual 
understanding, since the order of some steps is interchangeable, while others might not, and in model diagnostics, you might encounter different plots and values than the ones 
shown here, which require different treatments. 

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- ROADMAP -->
## Roadmap

- [ ] Power Transformation 
- [ ] Preconditions Check
- [ ] Assumptions Check
- [ ] Variable Selection
- [ ] Analysis of Covariance
- [ ] High Leverage Points Check
- [ ] Preconditions Recheck
- [ ] Assumptions Recheck
- [ ] Model Validation
- [ ] Summary Statistics
    - [ ] Histogram
    - [ ] Boxplot
    - [ ] Scatterplot
    - [ ] Correlation table

<p align="right">(<a href="#top">back to top</a>)</p>


## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your Feature Branch 
3. Commit your Changes
4. Push to the Branch 
5. Open a Pull Request

<p align="right">(<a href="#top">back to top</a>)</p>


## License

Distributed under the MIT License. See `LICENSE.txt` for more information.

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- CONTACT -->
## Contact

Sherry Xiaoman Lu - sherry.luxiaoman@gmail.com
[![LinkedIn][linkedin-shield]][linkedin-url]

Project Link: [https://github.com/SherryLuXM/STA302-Final-Project](https://github.com/SherryLuXM/STA302-Final-Project)

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- ACKNOWLEDGMENTS -->
## Acknowledgments

* This is the final project from STA302/1001 Methods of Data Analysis 1, taught by Katherine Daignault, taken at University of Toronto in the 2020-2021 Winter term. Professor
Daignault has provided invaluable training and support. 

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- MARKDOWN LINKS & IMAGES -->
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&colorB=555
[linkedin-url]: https://www.linkedin.com/in/sherry-l-633854132/


Changes
scheduler.py
might need to change attributes in GreedyAlgorithm to private


a1_starter_tests.py
"""Assignment 1 - Tests

CSC148, Winter 2021

This code is provided solely for the personal and private use of
students taking the CSC148 course at the University of Toronto.
Copying for purposes other than this use is expressly prohibited.
All forms of distribution of this code, whether as given or with
any changes, are expressly prohibited.

Authors: Diane Horton, Ian Berlott-Atwell, Jonathan Calver,
Sophia Huynh, Myriam Majedi, and Jaisie Sin.

All of the files in this directory and all subdirectories are:
Copyright (c) 2021 Diane Horton, Ian Berlott-Atwell, Jonathan Calver,
Sophia Huynh, Myriam Majedi, and Jaisie Sin.

===== Module Description =====

This module contains sample test cases that you can use to test your code.
These are a very incomplete set of test cases! We will be testing your code on
a much more thorough set of tests.

The self-test on MarkUs runs all of the tests below, along with a few others.
Make sure you run the self-test on MarkUs after submitting your code!

Once you have the entire program completed, that is, after Task 5, your
code should pass all of the tests we've provided.  As you develop your code,
test cases for parts that you haven't written yet will fail, of course.

But as you work through the earlier phases of the assignment, you can run the
individual tests below for each method as you complete it. We encourage you to
add further test cases in this file to improve your confidence in your code.

Tip: if you put your mouse inside a pytest function and right click, the "run"
menu will give you the option of running just that test function.
"""
import pytest
from typing import Dict
from distance_map import DistanceMap
from domain import Truck, Parcel, Fleet
from scheduler import GreedyScheduler
from container import PriorityQueue, _shorter
from experiment import SchedulingExperiment

# This variable is used in the special pytest test case defined by function
# test_experiment below.  The variable defines a single scheduling experiment
# test case to be run. It gives a unique identifier for the test case, and
# specifies both the configuration to use and the correct statistics to expect.
test_arguments = [
    ('1-small',
     {
         'depot_location': 'Toronto',
         'parcel_file': 'data/parcel-data-small.txt',
         'truck_file': 'data/truck-data-small.txt',
         'map_file': 'data/map-data.txt',
         'algorithm': 'greedy',
         'parcel_priority': 'volume',
         'parcel_order': 'non-decreasing',
         'truck_order': 'non-decreasing',
         'verbose': 'false'
     },
     {
         'fleet': 3,
         'unused_trucks': 0,
         'unused_space': 0,
         'avg_distance': 192.7,
         'avg_fullness': 100,
         'unscheduled': 0
     }),
    # You can add additional test cases here!
    # Write these in the format:
    # (<test_id>, <config dictionary>, <expected_stats dictionary>)
    # If you're adding multiple tests: remember to add a comma (,) after the
    # tuple!
    ('2-destination',
     {
         'depot_location': 'Toronto',
         'parcel_file': 'data/parcel-data-small.txt',
         'truck_file': 'data/truck-data-small.txt',
         'map_file': 'data/map-data.txt',
         'algorithm': 'greedy',
         'parcel_priority': 'destination',
         'parcel_order': 'non-increasing',
         'truck_order': 'non-decreasing',
         'verbose': 'false'
     },
     {
         'fleet': 3,
         'unused_trucks': 0,
         'unused_space': 0,
         'avg_distance': 192.7,
         'avg_fullness': 100,
         'unscheduled': 0
     }),
]


def test_distance_map_basic() -> None:
    """Test DistanceMap when a single distance is provided."""
    m = DistanceMap()
    assert m.distance('Montreal', 'Toronto') == -1
    m.add_distance('Montreal', 'Toronto', 4)
    assert m.distance('Montreal', 'Toronto') == 4


def test_distance_map_multiple() -> None:
    """Test DistanceMap with multiple distances provided."""
    m = DistanceMap()
    assert m.distance('Vancouver', 'Toronto') == -1
    assert m.distance('Edmonton', 'Toronto') == -1
    m.add_distance('Montreal', 'Toronto', 4)
    m.add_distance('Calgary', 'Toronto', 10)
    m.add_distance('Edmonton', 'Toronto', 11)
    m.add_distance('Vancouver', 'Toronto', 20)

    assert m.distance('Montreal', 'Toronto') == 4
    assert m.distance('Calgary', 'Toronto') == 10
    assert m.distance('Edmonton', 'Toronto') == 11
    assert m.distance('Vancouver', 'Toronto') == 20


def test_num_trucks_doctest() -> None:
    """Test the doctest provided for Fleet.num_trucks"""
    f = Fleet()
    assert f.num_trucks() == 0
    t1 = Truck(1423, 10, 'Toronto')
    t2 = Truck(1200, 10, 'Toronto')
    t3 = Truck(1300, 10, 'Toronto')
    t4 = Truck(1250, 5, 'Toronto')
    f.add_truck(t1)
    f.add_truck(t2)
    f.add_truck(t3)
    f.add_truck(t4)
    assert f.num_trucks() == 4


def test_num_nonempty_trucks_doctest() -> None:
    """Test the doctest provided for Fleet.num_nonempty_trucks"""
    f = Fleet()

    t1 = Truck(1423, 10, 'Toronto')
    f.add_truck(t1)
    p1 = Parcel(1, 5, 'Buffalo', 'Hamilton')
    assert t1.pack(p1) is True

    p2 = Parcel(2, 4, 'Toronto', 'Montreal')
    assert t1.pack(p2) is True
    assert t1.fullness() == 90.0

    t2 = Truck(5912, 20, 'Toronto')
    f.add_truck(t2)
    p3 = Parcel(3, 2, 'New York', 'Windsor')
    assert t2.pack(p3) is True
    assert t2.fullness() == 10.0

    t3 = Truck(1111, 50, 'Toronto')
    f.add_truck(t3)
    assert f.num_nonempty_trucks() == 2


def test_parcel_allocations_doctest() -> None:
    """Test the doctest provided for Fleet.parcel_allocations"""
    f = Fleet()
    t1 = Truck(1423, 10, 'Toronto')
    p1 = Parcel(27, 5, 'Toronto', 'Hamilton')
    p2 = Parcel(12, 5, 'Toronto', 'Hamilton')
    assert t1.pack(p1) is True
    assert t1.pack(p2) is True
    t2 = Truck(1333, 10, 'Toronto')
    p3 = Parcel(28, 5, 'Toronto', 'Hamilton')
    assert t2.pack(p3) is True
    f.add_truck(t1)
    f.add_truck(t2)
    assert f.parcel_allocations() == {1423: [27, 12], 1333: [28]}


def test_total_unused_space_doctest() -> None:
    """Test the doctest provided for Fleet.total_unused_space"""
    f = Fleet()
    assert f.total_unused_space() == 0

    t = Truck(1423, 1000, 'Toronto')
    p = Parcel(1, 5, 'Buffalo', 'Hamilton')
    assert t.pack(p) is True

    f.add_truck(t)
    assert f.total_unused_space() == 995


def test_average_fullness_doctest() -> None:
    """Test the doctest provided for Fleet.average_fullness"""
    f = Fleet()
    t = Truck(1423, 10, 'Toronto')
    p = Parcel(1, 5, 'Buffalo', 'Hamilton')
    assert t.pack(p) is True

    f.add_truck(t)
    assert f.average_fullness() == 50.0


def test_total_distance_travelled_doctest() -> None:
    """Test the doctest provided for Fleet.total_distance_travelled"""
    f = Fleet()
    t1 = Truck(1423, 10, 'Toronto')
    p1 = Parcel(1, 5, 'Toronto', 'Hamilton')
    assert t1.pack(p1) is True

    t2 = Truck(1333, 10, 'Toronto')
    p2 = Parcel(2, 5, 'Toronto', 'Hamilton')
    assert t2.pack(p2) is True

    m = DistanceMap()
    m.add_distance('Toronto', 'Hamilton', 9)
    f.add_truck(t1)
    f.add_truck(t2)
    assert f.total_distance_travelled(m) == 36


def test_average_distance_travelled_doctest() -> None:
    """Test the doctest provided for Fleet.average_distance_travelled"""
    f = Fleet()
    t1 = Truck(1423, 10, 'Toronto')
    p1 = Parcel(1, 5, 'Toronto', 'Hamilton')
    assert t1.pack(p1) is True

    t2 = Truck(1333, 10, 'Toronto')
    p2 = Parcel(2, 5, 'Toronto', 'Hamilton')
    assert t2.pack(p2) is True

    m = DistanceMap()
    m.add_distance('Toronto', 'Hamilton', 9)
    f.add_truck(t1)
    f.add_truck(t2)
    assert f.average_distance_travelled(m) == 18.0


def test_priority_queue_is_empty_doctest() -> None:
    """Test the doctest provided for PriorityQueue.is_empty"""
    pq = PriorityQueue(str.__lt__)
    assert pq.is_empty() is True

    pq.add('fred')
    assert pq.is_empty() is False


def test_priority_queue_add_remove_doctest() -> None:
    """Test the doctest provided for PriorityQueue.add and
    PriorityQueue.remove"""
    pq = PriorityQueue(_shorter)
    pq.add('fred')
    pq.add('arju')
    pq.add('monalisa')
    pq.add('hat')
    assert pq.remove() == 'hat'
    assert pq.remove() == 'fred'
    assert pq.remove() == 'arju'
    assert pq.remove() == 'monalisa'


def test_greedy_scheduler_example() -> None:
    """Test GreedyScheduler on the example provided."""
    p17 = Parcel(17, 25, 'York', 'Toronto')
    p21 = Parcel(21, 10, 'York', 'London')
    p13 = Parcel(13, 8, 'York', 'London')
    p42 = Parcel(42, 20, 'York', 'Toronto')
    p25 = Parcel(25, 15, 'York', 'Toronto')
    p61 = Parcel(61, 15, 'York', 'Hamilton')
    p76 = Parcel(76, 20, 'York', 'London')

    t1 = Truck(1, 40, 'York')
    t2 = Truck(2, 40, 'York')
    t3 = Truck(3, 25, 'York')

    f = Fleet()
    f.add_truck(t1)
    f.add_truck(t2)
    f.add_truck(t3)

    # We've left parcel_file, truck_file, and map_file empty in the config
    # dictionary below because you should *not* use these in your
    # GreedyScheduler. It is not responsible for reading data from these files.
    config = {'depot_location': 'York',
              'parcel_file': '',
              'truck_file': '',
              'map_file': '',
              'algorithm': 'greedy',
              'parcel_priority': 'destination',
              'parcel_order': 'non-increasing',
              'truck_order': 'non-increasing',
              'verbose': 'false'}

    scheduler = GreedyScheduler(config)
    unscheduled = scheduler.schedule([p17, p21, p13, p42, p25, p61, p76],
                                     [t1, t2, t3])

    assert unscheduled == [p76]

    truck_parcels = f.parcel_allocations()
    assert truck_parcels[1] == [17, 61]
    assert truck_parcels[2] == [42, 25]
    assert truck_parcels[3] == [21, 13]


################################################################################
# The test below uses pytest.mark.parametrize.
#
# This provides a way of running the same test code with different parameters
# without having to repeat the body multiple times.
#
# The line above the test_experiment method works as follows:
# @pytest.mark.parametrize('test_id, config, expected_stats', test_arguments)
#                           ^                                 ^
#                    These are the parameters        This is a list where each
#                    of test_experiment that         element is a tuple
#                    we're filling.                  containing values for those
#                                                    parameters.
#
# test_arguments is the variable defined near the top of this module.
#
# We have included one item in test_arguments for you. This is a tuple
# representing the following parameter configurations:
#     test_id        = '1-small'
#     config         = {'depot_location': 'Toronto', ...}
#     expected_stats = {'fleet': 3, ...}
#
# If you want to add additional test cases, you create a tuple with the same
# format and add it to the list.
#
# For more details, see:
# https://docs.pytest.org/en/stable/parametrize.html
#
# NOTE: if you get a "FileNotFoundError", try replacing the filename
# with the full path to the file (e.g., "C:\\Users\\David\\Documents\\...")
################################################################################
@pytest.mark.parametrize('stat', [
    'fleet', 'unused_trucks', 'unused_space', 'avg_distance', 'avg_fullness',
    'unscheduled'])
class TestExperiment:
    """
    Tests for SchedulingExperiment.run
    """

    @pytest.mark.parametrize('test_id, config, expected_stats', test_arguments)
    def test_experiment(self, test_id: str, config: Dict[str, str],
                        expected_stats: Dict[str, str], stat: str) -> None:
        """Run the SchedulingExperiment on the given config and expected_stats.
        Assert that the stat returned from the experiment matches
        expected_stats[stat].
        """
        experiment = SchedulingExperiment(config)
        results = experiment.run()

        # pytest.approx lets us use approximate values so we can avoid
        # failing a test case over very small differences in floating point
        # values. These can arise simply from doing mathematical operations in
        # a different order. [If you find this intriguing, take csc336!]

        # In this case, we're making sure our actual value is in the range
        # (expected - 1e-1, expected + 1e-1)
        expected = expected_stats[stat]
        actual = results[stat]
        assert actual == pytest.approx(expected, abs=1e-1)


if __name__ == '__main__':
    pytest.main(['a1_starter_tests.py'])


domain.py
"""Assignment 1 - Domain classes (Task 2)

CSC148, Winter 2021

This code is provided solely for the personal and private use of
students taking the CSC148 course at the University of Toronto.
Copying for purposes other than this use is expressly prohibited.
All forms of distribution of this code, whether as given or with
any changes, are expressly prohibited.

Authors: Diane Horton, Ian Berlott-Atwell, Jonathan Calver,
Sophia Huynh, Myriam Majedi, and Jaisie Sin.

All of the files in this directory and all subdirectories are:
Copyright (c) 2021 Diane Horton, Ian Berlott-Atwell, Jonathan Calver,
Sophia Huynh, Myriam Majedi, and Jaisie Sin.

===== Module Description =====

This module contains the classes required to represent the entities
in the simulation: Parcel, Truck and Fleet.
"""
from typing import List, Dict
from distance_map import DistanceMap


class Parcel:
    """Create a parcel. Each parcel has a unique ID, volume (measured
    in cc), and a source and destination.

    === Public Attributes ===
    id_: a parcel's unique ID
    volume: how much space a parcel takes up in cubic centimetres
    source: name of the city it came from
    destination: name of the city its being delivered to

    === Representation Invariants ===
    - volume > 0
    - destination is not the depot
    === Sample Usage ===
    >>> p = Parcel(1, 10, 'Toronto', 'Calgary')
    >>> p.id_ == 1
    True
    >>> p.source == 'Toronto'
    True
    >>> p.destination == 'Calgary'
    True
    """
    id_: int
    volume: int
    source: str
    destination: str

    def __init__(self, id_: int, volume: int, source: str, destination: str) \
            -> None:
        """Create an instance of a parcel."""

        self.id_ = id_
        self.volume = volume
        self.source = source
        self.destination = destination


class Truck:
    """Create an instance of a truck. Each truck has a unique ID,
    volume capacity, and a Route.

    === Public Attributes ===
    id_: a truck's ID number
    volume_capacity: the maximum amount of volume a truck can carry.
    depot: a Truck's initial position.
    stored: how much volume is stored onto the Truck.
    route: an ordered List of city names that a truck is supposed to go through.
    parcels: the parcels allocated for the truck to deliver.
    === Representation Invariants ===
    - 0 <= stored <= volume_capacity
    - volume_capacity > 0
    - route[0] == depot

    === Sample Usage ===
    >>> t = Truck(1200, 10, 'Toronto')
    >>> t.id_ == 1200
    True
    >>> t.volume_capacity == 10
    True
    >>> p1 = Parcel(1, 2, 'Toronto', 'Ottawa')
    >>> p2 = Parcel(2, 3, 'Toronto', 'Ottawa')
    >>> t.pack(p1)
    True
    >>> t.pack(p2)
    True
    >>> t.fullness()
    50.0
    """
    id_: int
    volume_capacity: int
    depot: str
    stored: int
    route: List[str]
    parcels: List[Parcel]

    def __init__(self, id_: int, volume_capacity: int, depot: str) -> None:
        """Create a Truck. A Truck will always initially be empty and will
        always start at the depot."""

        self.id_ = id_
        self.volume_capacity = volume_capacity
        self.depot = depot
        self.stored = 0
        self.route = [self.depot]
        self.parcels = []

    def packable(self, parcel: Parcel) -> bool:
        """
        return True if it is possible to pack <parcel> onto the truck.
        >>> t = Truck(1000, 10, 'Toronto')
        >>> p1 = Parcel(1, 5, 'Toronto', 'Ottawa')
        >>> p2 = Parcel(2, 6, 'Toronto', 'Calgary')
        >>> t.packable(p1)
        True
        >>> t.packable(p2)
        True
        >>> t.pack(p1)
        True
        >>> t.pack(p2)
        False
        >>> t.route[-1]
        'Ottawa'
        >>> t.fullness()
        50.0
        """
        return parcel.volume + self.stored <= self.volume_capacity

    def pack(self, parcel: Parcel) -> bool:
        """Pack the Truck with a Parcel, return True if it has been
        successfully packed. Return False if stored exceeds
        volume_capacity.
        Add the parcel's destination to the end of the Truck's route unless the
        LAST item is equal to the destination.
        >>> t = Truck(1000, 10, 'Toronto')
        >>> p1 = Parcel(1, 5, 'Toronto', 'Ottawa')
        >>> p2 = Parcel(2, 6, 'Toronto', 'Calgary')
        >>> t.pack(p1)
        True
        >>> t.pack(p2)
        False
        >>> t.route[-1]
        'Ottawa'
        >>> t.fullness()
        50.0
        """
        # First, check if there's enough space to fit the parcel.
        if self.packable(parcel):
            # Add the parcel to the Truck.
            self.stored += parcel.volume
            self.parcels.append(parcel)
            # Don't modify route if the last item is the same as the
            # parcel's destination.
            if self.route[-1] != parcel.destination:
                self.route.append(parcel.destination)
            return True
        # At this point we know the parcel doesn't fit.
        return False

    def fullness(self) -> float:
        """Return the percentage of a Truck's fullness.
        >>> t = Truck(1000, 10, 'Toronto')
        >>> p1 = Parcel(1, 5, 'Toronto', 'Ottawa')
        >>> p2 = Parcel(1, 5, 'Toronto', 'Calgary')
        >>> t.pack(p1)
        True
        >>> t.pack(p2)
        True
        >>> t.fullness()
        100.0
        """
        return (100 * self.stored) / self.volume_capacity

    def distance(self, dmap: DistanceMap) -> int:
        """calculate the distance travelled by this truck, with data from <dmap>

        >>> t = Truck(1333, 10, 'Toronto')
        >>> p = Parcel(2, 5, 'Toronto', 'Hamilton')
        >>> t.pack(p)
        True
        >>> m = DistanceMap()
        >>> m.add_distance('Toronto', 'Hamilton', 9)
        >>> t.distance(m)
        18
        """
        i = 0
        d_total = 0
        while i < len(self.route) - 1:
            dist = dmap.distance(self.route[i], self.route[i + 1])
            if dist > 0:
                d_total += dmap.distance(self.route[i], self.route[i + 1])
            i += 1
        # the truck needs to return to depot at the end of route
        final = dmap.distance(self.route[i], self.depot)
        if self.route[i] != self.depot and final > 0:
            d_total += final
        return d_total


class Fleet:
    """ A fleet of trucks for making deliveries.
    ===== Public Attributes =====
    trucks:
      List of all Truck objects in this fleet.
    """
    trucks: List[Truck]

    def __init__(self) -> None:
        """Create a Fleet with no trucks.

        >>> f = Fleet()
        >>> f.num_trucks()
        0
        """
        self.trucks = []

    def add_truck(self, truck: Truck) -> None:
        """Add <truck> to this fleet.

        Precondition: No truck with the same ID as <truck> has already been
        added to this Fleet.
        >>> f = Fleet()
        >>> t = Truck(1423, 1000, 'Toronto')
        >>> f.add_truck(t)
        >>> f.num_trucks()
        1
        """
        self.trucks.append(truck)

    # We will not test the format of the string that you return -- it is up
    # to you.
    def __str__(self) -> str:
        """Produce a string representation of this fleet

        >>> f = Fleet()
        >>> t1 = Truck(1423, 10, 'Toronto')
        >>> f.add_truck(t1)
        >>> print(f)
        Information of all trucks in the fleet:
        ID: 1423; Depot: Toronto; Storage: 0/10; Route: ['Toronto']
        Total: 1 truck(s)
        """
        info = 'Information of all trucks in the fleet:'
        for truck in self.trucks:
            info = info + '\n' + f'ID: {truck.id_}; Depot: {truck.depot}; ' \
                                 f'Storage: {truck.stored}/' \
                                 f'{truck.volume_capacity}; ' \
                                 f'Route: {truck.route}'

        info += f'\nTotal: {self.num_trucks()} truck(s)'
        return info

    def num_trucks(self) -> int:
        """Return the number of trucks in this fleet.

        >>> f = Fleet()
        >>> t1 = Truck(1423, 10, 'Toronto')
        >>> f.add_truck(t1)
        >>> f.num_trucks()
        1
        """
        return len(self.trucks)

    def num_nonempty_trucks(self) -> int:
        """Return the number of non-empty trucks in this fleet.

        >>> f = Fleet()
        >>> t1 = Truck(1423, 10, 'Toronto')
        >>> f.add_truck(t1)
        >>> p1 = Parcel(1, 5, 'Buffalo', 'Hamilton')
        >>> t1.pack(p1)
        True
        >>> p2 = Parcel(2, 4, 'Toronto', 'Montreal')
        >>> t1.pack(p2)
        True
        >>> t1.fullness()
        90.0
        >>> t2 = Truck(5912, 20, 'Toronto')
        >>> f.add_truck(t2)
        >>> p3 = Parcel(3, 2, 'New York', 'Windsor')
        >>> t2.pack(p3)
        True
        >>> t2.fullness()
        10.0
        >>> t3 = Truck(1111, 50, 'Toronto')
        >>> f.add_truck(t3)
        >>> f.num_nonempty_trucks()
        2
        """
        count = 0
        for truck in self.trucks:
            if truck.stored != 0:
                count += 1
        return count

    def parcel_allocations(self) -> Dict[int, List[int]]:
        """Return a dictionary in which each key is the ID of a truck in this
        fleet and its value is a list of the IDs of the parcels packed onto it,
        in the order in which they were packed.

        >>> f = Fleet()
        >>> t1 = Truck(1423, 10, 'Toronto')
        >>> p1 = Parcel(27, 5, 'Toronto', 'Hamilton')
        >>> p2 = Parcel(12, 5, 'Toronto', 'Hamilton')
        >>> t1.pack(p1)
        True
        >>> t1.pack(p2)
        True
        >>> t2 = Truck(1333, 10, 'Toronto')
        >>> p3 = Parcel(28, 5, 'Toronto', 'Hamilton')
        >>> t2.pack(p3)
        True
        >>> f.add_truck(t1)
        >>> f.add_truck(t2)
        >>> f.parcel_allocations() == {1423: [27, 12], 1333: [28]}
        True
        """
        allocations = {}
        for truck in self.trucks:
            allocations[truck.id_] = []
            for parcel in truck.parcels:
                allocations[truck.id_].append(parcel.id_)
        return allocations

    def total_unused_space(self) -> int:
        """Return the total unused space, summed over all non-empty trucks in
        the fleet.
        If there are no non-empty trucks in the fleet, return 0.

        >>> f = Fleet()
        >>> f.total_unused_space()
        0
        >>> t = Truck(1423, 1000, 'Toronto')
        >>> p = Parcel(1, 5, 'Buffalo', 'Hamilton')
        >>> t.pack(p)
        True
        >>> f.add_truck(t)
        >>> f.total_unused_space()
        995
        """
        unused_space = 0
        for truck in self.trucks:
            stored = truck.stored
            if stored != 0:
                unused_space += (truck.volume_capacity - stored)
        return unused_space

    def _total_fullness(self) -> float:
        """Return the sum of truck.fullness() for each non-empty truck in the
        fleet. If there are no non-empty trucks, return 0.

        >>> f = Fleet()
        >>> f._total_fullness() == 0.0
        True
        >>> t = Truck(1423, 10, 'Toronto')
        >>> f.add_truck(t)
        >>> f._total_fullness() == 0.0
        True
        >>> p = Parcel(1, 5, 'Buffalo', 'Hamilton')
        >>> t.pack(p)
        True
        >>> f._total_fullness()
        50.0
        """
        flt_fullness = 0
        for truck in self.trucks:
            flt_fullness += truck.fullness()
        return flt_fullness

    def average_fullness(self) -> float:
        """Return the average percent fullness of all non-empty trucks in the
        fleet.

        Precondition: At least one truck is non-empty.

        >>> f = Fleet()
        >>> t = Truck(1423, 10, 'Toronto')
        >>> p = Parcel(1, 5, 'Buffalo', 'Hamilton')
        >>> t.pack(p)
        True
        >>> f.add_truck(t)
        >>> f.average_fullness()
        50.0
        """
        return self._total_fullness() / self.num_nonempty_trucks()

    def total_distance_travelled(self, dmap: DistanceMap) -> int:
        """Return the total distance travelled by the trucks in this fleet,
        according to the distances in <dmap>.

        Precondition: <dmap> contains all distances required to compute the
                      average distance travelled.

        >>> f = Fleet()
        >>> t1 = Truck(1423, 10, 'Toronto')
        >>> p1 = Parcel(1, 5, 'Toronto', 'Hamilton')
        >>> t1.pack(p1)
        True
        >>> t2 = Truck(1333, 10, 'Toronto')
        >>> p2 = Parcel(2, 5, 'Toronto', 'Hamilton')
        >>> t2.pack(p2)
        True
        >>> m = DistanceMap()
        >>> m.add_distance('Toronto', 'Hamilton', 9)
        >>> f.add_truck(t1)
        >>> f.add_truck(t2)
        >>> f.total_distance_travelled(m)
        36
        """
        total_d = 0
        for truck in self.trucks:
            dist = truck.distance(dmap)
            if dist > 0:
                total_d += dist
        return total_d

    def average_distance_travelled(self, dmap: DistanceMap) -> float:
        """Return the average distance travelled by the trucks in this fleet,
        according to the distances in <dmap>.
        Include in the average only trucks that have actually travelled some
        non-zero distance.

        === Preconditions ===
        - <dmap> contains all distances required to compute the average
          distance travelled.
        - At least one truck has travelled a non-zero distance.

        >>> f = Fleet()
        >>> t1 = Truck(1423, 10, 'Toronto')
        >>> p1 = Parcel(1, 5, 'Toronto', 'Hamilton')
        >>> t1.pack(p1)
        True
        >>> t2 = Truck(1333, 10, 'Toronto')
        >>> p2 = Parcel(2, 5, 'Toronto', 'Hamilton')
        >>> t2.pack(p2)
        True
        >>> m = DistanceMap()
        >>> m.add_distance('Toronto', 'Hamilton', 9)
        >>> f.add_truck(t1)
        >>> f.add_truck(t2)
        >>> f.average_distance_travelled(m)
        18.0
        """
        total = self.total_distance_travelled(dmap)
        if total == 0:
            return 0
        count = 0
        for truck in self.trucks:
            # Only add trucks that have travelled a non-zero distance.
            if truck.distance(dmap) > 0:
                count += 1
        return total / count


if __name__ == '__main__':
    import python_ta
    python_ta.check_all(config={
        'allowed-import-modules': ['doctest', 'python_ta', 'typing',
                                   'distance_map'],
        'disable': ['E1136'],
        'max-attributes': 15,
    })
    import doctest
    doctest.testmod()
